test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -66,7 +67,22 @@ class test_ucp_sockaddr : public ucp_test {
         ASSERT_UCS_OK(status);
     }
 
-    void tag_send_recv(entity& from, entity& to) {
+    void wait(ucp_worker_h worker, void *request, bool wakeup)
+    {
+        if (wakeup) {
```

Comment:
can be written shorter, also isn't progressing all time time causing the test to "hang"?
```
do {
     if (wait) {
        ucp_worker_wait(worker);
     } else {
        sched_yield();
     }
     while (progress());
} while (!ucp_request_is_completed(request));
```

Reply:
can you please clarify the question?

Reply:
you mentioned before that doing progress in a loop without sleep/yield causes tarvation of progress thread. but it is done here in case of wakeup==false

Reply:
the hang happened on the wakeup=true case. when the connection wasn't established yet when the sender was waiting for a completion on the send but didn't receive completion on the wireup message yet since the receiver probably didnt get the wireup connection request yet. 
this is why i wait for a connection establishment event first (answering the question from the other comment here as well)

Reply:
ok,  but now the sender is going to sleep by doing worker_wait(), we should be able to remove the code which waits for wireup

Reply:
the code that "waits" for wireup isnt in the 'wait' function. it's the short_progress_loop() in connect_and_send_recv() before even starting send-recv. 
if i remove it, the test may hang since send-recv will start before connection is established (since no one progressed the server side to make sure it reveived the wireup connection request).

Reply:
but according to UCP APi definition, the user is allowed to start sending immediately after client ep is created. only the receive should wait until server ep is created.

Reply:
ok, then inside connect_and_send_recv(), need this?
```
if (wakeup) {
            ucp_worker_wait(receiver().worker());
            while (progress());       --> while (receiver().progress()); 
            short_progress_loop();    --> remove

            tag_send_recv(sender(), receiver() ,true);
            tag_send_recv(receiver(), sender(), true);
        }
```

how can we make sure though that the server is connected (wireup wise)? and not only that it had the connection establishment event (client-server wise)?

Reply:
no need for the above. 
Need to progress only as part of waiting for send or receive request completion, or waiting for server ep to appear.
In this specific test, seems there is **no need** to make sure server is connected (or waiting for server ep to appear), because receive is done on the worker (tag api).